panther_manager/test/test_safety_behavior_tree.cpp

```diff
@@ -0,0 +1,422 @@
+}
+
+void TestSafetyBehaviorTree::PublishBatteryState(
+  const std::uint8_t status, const std::uint8_t health, const float temperature)
+{
+  BatteryStateMsg msg;
+  msg.power_supply_status = status;
+  msg.power_supply_health = health;
+  msg.temperature = temperature;
+  battery_state_pub_->publish(msg);
+}
+
+void TestSafetyBehaviorTree::PublishIOState(const bool fan, const bool aux_power)
+{
+  IOStateMsg msg;
+  msg.fan = fan;
+  msg.aux_power = aux_power;
+  io_state_pub_->publish(msg);
+}
+
+void TestSafetyBehaviorTree::PublishDriverState(const float temperature)
+{
+  DriverStateMsg msg;
+  msg.front.temperature = temperature;
+  msg.rear.temperature = temperature;
+  driver_state_pub_->publish(msg);
+}
+
+void TestSafetyBehaviorTree::PublishSystemStatus(const float cpu_temp)
+{
+  SystemStatusMsg msg;
+  msg.cpu_temp = cpu_temp;
+  system_status_pub_->publish(msg);
+}
+
+void TestSafetyBehaviorTree::FanServerCB(
+  const SetBoolSrv::Request::SharedPtr & request, SetBoolSrv::Response::SharedPtr response)
+{
+  fan_state_ = request->data;
+  PublishIOState(fan_state_, aux_power_state_);
+  response->success = true;
+}
+
+void TestSafetyBehaviorTree::AUXPowerServerCB(
+  const SetBoolSrv::Request::SharedPtr & request, SetBoolSrv::Response::SharedPtr response)
+{
+  aux_power_state_ = request->data;
+  PublishIOState(fan_state_, aux_power_state_);
+  response->success = true;
+}
+
+void TestSafetyBehaviorTree::EStopTriggerSrvCB(
+  const TriggerSrv::Request::SharedPtr & /*request*/, TriggerSrv::Response::SharedPtr response)
+{
+  e_stop_triggered_ = true;
+  response->success = true;
+}
+
+TEST_F(TestSafetyBehaviorTree, TurnOnFanAtStartup)
+{
+  safety_manager_node_->Initialize();
+  PublishDefaultStates();
+
+  SpinWhileRunning();
+  EXPECT_TRUE(fan_state_);
+
+  // wait for automatic turn off of the fan
+  std::this_thread::sleep_for(
+    std::chrono::milliseconds(static_cast<unsigned>(kFanTurnOffTimeout * 1000)));
+  SpinWhileRunning();
+  EXPECT_FALSE(fan_state_);
+}
+
+TEST_F(TestSafetyBehaviorTree, TurnOnFanCpuTemp)
+{
+  const float high_cpu_temp = 70.1;
+  const float edge_cpu_temp = 59.9;
+
+  safety_manager_node_->Initialize();
+  PublishDefaultStates();
+
+  SpinWhileRunning();
+  EXPECT_TRUE(fan_state_);
+
+  // wait for automatic turn off of the fan
+  std::this_thread::sleep_for(
+    std::chrono::milliseconds(static_cast<unsigned>(kFanTurnOffTimeout * 1000)));
+  SpinWhileRunning();
+  EXPECT_FALSE(fan_state_);
+
+  PublishSystemStatus((2.0 * high_cpu_temp));
+  std::this_thread::sleep_for(
+    std::chrono::milliseconds(static_cast<unsigned>(kFanTurnOffTimeout * 1000)));
+  SpinWhileRunning();
+
+  EXPECT_TRUE(fan_state_);
+
+  PublishSystemStatus(3.0 * edge_cpu_temp - 2.0 * high_cpu_temp);
+  // wait for automatic turn off of the fan
+  std::this_thread::sleep_for(
+    std::chrono::milliseconds(static_cast<unsigned>(kFanTurnOffTimeout * 1000)));
+  SpinWhileRunning();
+  EXPECT_FALSE(fan_state_);
```

Comment:
What stands behind these formulas? Could you provide a brief explanation?
`PublishSystemStatus((2.0 * high_cpu_temp));`
`PublishSystemStatus(3.0 * edge_cpu_temp - 2.0 * high_cpu_temp);`

Reply:
That was required by moving average. I changed `window_len` params to simplify this.